main.cpp

```diff
@@ -1,23 +1,134 @@
 #include <bits/stdc++.h>
+
+#define nol nums_of_lfs
+
 using namespace std;
 
-int n;
-vector <char> cars;
+struct lf
+{
+    int indx, val;
+};
 
-int main()
+bool operator<(lf l, lf r)
+    { return (l.val != r.val ? l.val < r.val : l.indx < r.indx); }
+
+struct Node
+{
+    set <lf> lfs;
+};
+
+struct Query
 {
-	ios_base::sync_with_stdio(false);
-	cin.tie(nullptr);
+    int type, a, b;
+};
+
+int n, q, nums_of_lfs;
 
-	cin >> n;
-	cars.resize(n - 1);
-	for(int i = 0; i < cars.size(); ++i)
-		cin >> cars[i];
+vector <int> arr;
+vector <Node> tree;
 
-	return 0;
+void FillTree(int lvl_l = nol / 2, int node = nol / 2)
+{
+    tree[node].lfs = tree[node * 2].lfs;
+    tree[node].lfs.insert(
+            tree[(node * 2 ) + 1].lfs.begin(), 
+            tree[(node * 2) + 1].lfs.end()
+            );
+
+    if (node == 1)
+    {
+        return;
+    }
+    else
+    {
+        FillTree(
+                node - lvl_l == lvl_l - 1 ? lvl_l / 2 : lvl_l, 
+                node - lvl_l == lvl_l - 1 ? lvl_l / 2 : node + 1\
+                );
+    }
```

Comment:
use insert instead of FillingTree function, will take less code and will be faster.